x/lightclient/keeper/hook_listener.go

```diff
@@ -0,0 +1,120 @@
+func (k Keeper) RollappHooks() rollapptypes.RollappHooks {
+	return rollappHook{k: k}
+}
+
+// AfterUpdateState is called after a state update is made to a rollapp.
+// This hook checks if the rollapp has a canonical IBC light client and if the Consensus state is compatible with the state update
+// and punishes the sequencer if it is not
+func (hook rollappHook) AfterUpdateState(
+	ctx sdk.Context,
+	rollappId string,
+	stateInfo *rollapptypes.StateInfo,
+	isFirstStateUpdate bool,
+	previousStateHasTimestamp bool,
+) error {
+	canonicalClient, found := hook.k.GetCanonicalClient(ctx, rollappId)
+	if !found {
+		canonicalClient = hook.k.GetProspectiveCanonicalClient(ctx, rollappId, stateInfo)
+		if canonicalClient != "" {
+			hook.k.SetCanonicalClient(ctx, rollappId, canonicalClient)
+		}
+		return nil
+	}
+	bds := stateInfo.GetBDs()
+	for i, bd := range bds.GetBD() {
+		// Check if any optimistic updates were made for the given height
+		tmHeaderSigner, found := hook.k.GetConsensusStateSigner(ctx, canonicalClient, bd.GetHeight())
+		if !found {
+			continue
+		}
+		height := ibcclienttypes.NewHeight(1, bd.GetHeight())
+		consensusState, _ := hook.k.ibcClientKeeper.GetClientConsensusState(ctx, canonicalClient, height)
+		// Cast consensus state to tendermint consensus state - we need this to check the state root and timestamp and nextValHash
+		tmConsensusState, ok := consensusState.(*ibctm.ConsensusState)
+		if !ok {
+			return nil
+		}
+		ibcState := types.IBCState{
+			Root:               tmConsensusState.GetRoot().GetHash(),
+			ValidatorsHash:     tmHeaderSigner,
+			NextValidatorsHash: tmConsensusState.NextValidatorsHash,
+			Timestamp:          time.Unix(0, int64(tmConsensusState.GetTimestamp())),
+		}
+		sequencerPk, err := hook.k.GetSeqeuncerHash(ctx, stateInfo.Sequencer)
+		if err != nil {
+			return err
+		}
+		rollappState := types.RollappState{
+			BlockSequencer:  sequencerPk,
+			BlockDescriptor: bd,
+		}
+		// check if bd for next block exists in same state info
+		if i+1 < len(bds.GetBD()) {
+			rollappState.NextBlockSequencer = sequencerPk
+			rollappState.NextBlockDescriptor = bds.GetBD()[i+1]
+		}
+		err = types.CheckCompatibility(ibcState, rollappState)
+		if err != nil {
+			// Only require timestamp on BD if first ever update, or the previous update had BD
+			if errors.Is(err, types.ErrTimestampNotFound) && !isFirstStateUpdate && !previousStateHasTimestamp {
+				continue
+			}
+			// The BD for (h+1) is missing, cannot verify if the nextvalhash matches
+			if errors.Is(err, types.ErrNextBlockDescriptorMissing) {
+				return err
+			}
+			// If the state is not compatible,
+			// Take this state update as source of truth over the IBC update
+			// Punish the block proposer of the IBC signed header
+			sequencerAddr, err := getAddress(tmHeaderSigner)
+			if err != nil {
+				return err
+			}
+			err = hook.k.sequencerKeeper.JailSequencerOnFraud(ctx, sequencerAddr)
+			if err != nil {
+				return err
+			}
```

Comment:
instead of calling sequencer keeper directly, we need to do a general fraud call on the rollapp keeper
you can pass the sequencer as argument to that

Reply:
check if this works https://github.com/dymensionxyz/dymension/blob/31156de13f071d2f333694c8158a07513553dcec/x/rollapp/keeper/fraud_handler.go#L18-L19